apps/next/app/api/deploy/up/route.ts

```diff
@@ -1,11 +1,208 @@
+import { DIR_FOR_CLIENT_DEPLOYMENTS } from "@/app/server/util/constants";
+import { NixpackPlan } from "@/types/deployment";
 import { clerkClient } from "@clerk/nextjs";
 import { type NextRequest } from "next/server";
-import { spawn } from "node:child_process";
-import { createWriteStream, existsSync, mkdirSync } from "node:fs";
-import { Writable } from "node:stream";
+import { exec as execCallbackBased, spawn } from "node:child_process";
+import {
+  createWriteStream,
+  existsSync,
+  mkdirSync,
+  writeFileSync,
+} from "node:fs";
+import { Readable, Writable } from "node:stream";
+import { promisify } from "node:util";
+import { dirSync } from "tmp";
 
 export const dynamic = "force-dynamic";
 
+const exec = promisify(execCallbackBased);
+const org = process.env.DOCKERHUB_ORG;
+const repo = process.env.DOCKERHUB_REPO;
+const builderLongName = process.env.CLOUD_BUILDER_LONG_NAME || "";
+const builderName = process.env.CLOUD_BUILDER_NAME || "";
+
+// This code below is heavily based on:
+// https://nextjs.org/docs/app/building-your-application/routing/route-handlers#streaming
+
+// https://developer.mozilla.org/docs/Web/API/ReadableStream#convert_async_iterator_to_stream
+function iteratorToStream(iterator: AsyncGenerator<Uint8Array, void, unknown>) {
+  return new ReadableStream({
+    async pull(controller) {
+      const { value, done } = await iterator.next();
+
+      if (done) {
+        controller.close();
+      } else {
+        controller.enqueue(value);
+      }
+    },
+  });
+}
+
+function sleep(time: number) {
+  return new Promise((resolve) => {
+    setTimeout(resolve, time);
+  });
+}
+
+const encoder = new TextEncoder();
+
+// Thanks ChatGPT!
+async function* streamIterator(stderr: Readable) {
+  for await (const chunk of stderr) {
+    yield chunk.toString();
+  }
+}
+
+async function* makeIterator(buildTag: string, fileName: string) {
+  const { name: tempDirName } = dirSync({
+    tmpdir: DIR_FOR_CLIENT_DEPLOYMENTS,
+    name: buildTag,
+  });
+  const extractionStream = spawn("tar", ["xzfv", fileName, "-C", tempDirName]);
+  const extractionPromise = new Promise<void>((resolve, reject) => {
+    extractionStream.on("error", (error) => {
+      console.error("Error: ", error);
+      reject(error);
+    });
+    extractionStream.on("exit", (code) => {
+      if (code !== 0) {
+        reject(new Error(`Tarball extraction failed with code ${code}`));
+        return;
+      }
+
+      resolve();
+    });
+  });
+
+  let isFirstDataFromExtraction = true;
+  for await (const data of streamIterator(extractionStream.stderr)) {
+    if (isFirstDataFromExtraction) {
+      isFirstDataFromExtraction = false;
+      console.log(`***** METAL: Extract files for ${buildTag} *****`);
+    }
+
+    yield encoder.encode(`${data}`);
+  }
+
+  await extractionPromise;
+
+  yield encoder.encode(
+    `[<metal>]Deployment started. Tag is ${buildTag}.[</metal>]`
```

Comment:
did you try using `chalk` here? Might simplify things compared to the system of `<metal>` tags

Reply:
Done.